frontend/lib/types/DomainType.cpp

```diff
@@ -69,34 +71,61 @@ DomainType::getDomainType(Context* context, ID id, UniqueString name,
 
 const DomainType*
 DomainType::getGenericDomainType(Context* context) {
-  auto name = UniqueString::get(context, "_domain");
   auto id = getDomainID(context);
+  auto name = id.symbolName(context);
   SubstitutionsMap subs;
   const DomainType* instantiatedFrom = nullptr;
   return getDomainType(context, id, name, instantiatedFrom, subs).get();
 }
 
 const DomainType*
 DomainType::getRectangularType(Context* context,
+                               const QualifiedType& instance,
                                const QualifiedType& rank,
                                const QualifiedType& idxType,
-                               const QualifiedType& stridable) {
+                               const QualifiedType& strides) {
+  auto genericDomain = getGenericDomainType(context);
+
   SubstitutionsMap subs;
+  CHPL_ASSERT(rank.isParam() && rank.param()->isIntParam());
   subs.emplace(ID(UniqueString(), 0, 0), rank);
+  CHPL_ASSERT(idxType.isType());
   subs.emplace(ID(UniqueString(), 1, 0), idxType);
-  subs.emplace(ID(UniqueString(), 2, 0), stridable);
+  CHPL_ASSERT(strides.isParam() && strides.param()->isEnumParam() &&
+              strides.param()->toEnumParam()->value().id.symbolPath() ==
+                  "ChapelRange.strideKind");
+  subs.emplace(ID(UniqueString(), 2, 0), strides);
+
+
+  // Add substitution for _instance field
+  auto& rf = fieldsForTypeDecl(context, genericDomain,
+                               resolution::DefaultsPolicy::IGNORE_DEFAULTS,
+                               /* syntaxOnly */ true);
+  ID instanceFieldId;
+  for (int i = 0; i < rf.numFields(); i++) {
+    if (rf.fieldName(i) == USTR("_instance")) {
+      instanceFieldId = rf.fieldDeclId(i);
+      break;
+    }
+  }
+  if (instanceFieldId.isEmpty()) {
+    CHPL_ASSERT(isMissingBundledRecordType(context, genericDomain->id()));
+    instanceFieldId = ID(USTR("_instance"), 0, 0);
+  }
+  subs.emplace(instanceFieldId, instance);
```

Comment:
Remind me -- is the reason we need this that we want the `Domain` type not to be generic (and it's generic if we don't set `_instance`?)

Reply:
Yes, as well as IIRC preserving the `rank` etc info contained in `_instance` for successive invocations of `ctFromSubs`.